### PR TITLE
Intentions ACL enforcement updates

### PR DESCRIPTION
**Pull request by mkeeler**
This PR also contains the necessary OSS changes to support namespaces with Consul Enterprise.

### CODE REVIEW

File: agent/consul/intention_endpoint.go [hunk 1]
```diff
@@ -25,6 +25,17 @@ type Intention struct {
...
+	if _, ixn, err := state.IntentionGet(nil, id); err != nil {
+		return false, err
+	} else if ixn != nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
 // Apply creates or updates an intention in the data store.
```
Comment (mkcp) [inline agent/consul/intention_endpoint.go]: `Apply` is a pretty considerable fn here -- I would recommend expanding the doccomment to describe its expected usage, side effects, and gotchas/footguns that might catch future folks by surprise if there are any.

File: agent/consul/intention_endpoint.go [hunk 2]
```diff
@@ -240,10 +240,18 @@ func (s *Intention) Match(
 	}
 
 	if rule != nil {
-		// We go through each entry and test the destination to check if it
-		// matches.
+		var authzContext acl.AuthorizerContext
+		// Go through each entry to ensure we have intention:read for the resource.
+
+		// TODO - should we do this instead of filtering the result set? This will only allow
```
Comment (mkcp) [inline agent/consul/intention_endpoint.go]: Nonblocking: but gonna ping @banks for visibility and feedback on this

File: agent/structs/intention_oss.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+// +build !consulent
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/structs/intention_oss.go]: This is a Go build tag. It prevents this file from being included in builds where the `consulent` tag is passed to the compiler. Which is basically for Consul Enterprise

[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/structs/intention_oss.go]: There will be a corresponding enterprise only file for implementations of these functions that do more meaningful thing. So essentially we are implementing some hooks for where we can add enterprise features.

File: agent/consul/intention_endpoint_test.go [hunk 4]
```diff
@@ -389,6 +389,325 @@ service "foo" {
...
+		doIntentionDelete(t, wildcardPrefixToken.SecretID, true)
+	})
+
+	t.Run("allow-delete", func(t *testing.T) {
+		// tests that a token with all the required privileges can delete
+		// intentions with a wildcard destination
+		doIntentionDelete(t, writeToken.SecretID, false)
+	})
+}
+
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/consul/intention_endpoint_test.go]: You can parallelize subtests (by calling `t.Parallel()` within the func passed to `t.Run`) but in general our team has found this to be difficult to maintain. In many cases it would be fine (and we have parallelized them) but in many others you would have to take extra care to ensure that any resources setup outside of the parallelized subtests live for as long as all the other tests which would involve wait groups and such for many of our tests (any of those that defer calls to shutdown a test agent or remove a directory).

So our rule of thumb in general is to not parallelize subtest invocations.

Additionally for these tests, the "subtest" is really just meant to separate different steps of a test to make pin pointing failures easier. In these tests I have also put in logic so that subtest dependencies are handled and we will skip subtests that require a previous subtest completely successfully.

This is quite odd, but in reality this test along with most of our test suite are really integration level tests. There is no way currently to test an RPC endpoint outside of spinning up a server running raft along with everything else. So this test could be broken into 15 other tests that start a server, wait for various conditions and then do their singular check and finish. However our test sweet becomes more flaky (and takes longer to execute as we spin up more test servers/agents). So writing the test in this manner: 1 test server + subtests + subtest skipping based on dependencies is to allow pinpointing problems when a test fails easier *and* preventing many test server startups to reduce testing time and prevent flakiness.

File: agent/consul/intention_endpoint.go (hunk 2 above)
Comment (mkeeler) [reply agent/consul/intention_endpoint.go]: Yeah, the behavior here is odd but not insecure and for our own purposes it will work perfectly fine. Basically if you have permissions to read an intention by either the source side or destination but not both, and you attempt a `Intention.Match` rpc with your request setup to look for intentions on the side that you do not have permissions on, then you will get a permission denied.

In reality this seems fine as you can just request it for the correct side of things and the real use case for this is for Connect proxies to be able to query for intentions regarding themselves as destinations so that they can enforce them. And unless you go out of your way to mess up your ACLs, (like granting `service:write` to register the service/proxy but also having an explicit deny rule for intentions) then that destination service will have the necessary permissions.

Changing to using an acl filter on a result set instead shouldn't alter functionality in any meaningful way for any reasonable use-case. But I could argue that

File: agent/consul/intention_endpoint.go (hunk 1 above)
Comment (mkeeler) [reply agent/consul/intention_endpoint.go]: Unfortunately the doc says exactly what it does. The entire body here is just how we go about doing it. I thought about refactoring this into a single switch statement on the Op field of the RPC request that called one of 3 other funcs but didn't do it because it would be a larger change and unnecessary for the work I am doing.

The logic in here is weird because ACL enforcement and request validation are done a bit differently depending on the Op. Maybe I will refactor into 3 funcs.

File: acl/policy_authorizer.go [hunk 5]
```diff
@@ -340,6 +340,100 @@ func newPolicyAuthorizerFromRules(rules *PolicyRules, ent *Config) (Authorizer,
 	return p, nil
 }
 
+func anyAllowed(tree *radix.Tree, enforceFn func(raw interface{}, prefixOnly bool) EnforcementDecision) EnforcementDecision {
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply acl/policy_authorizer.go]: Going to create a type definition for this func type and add some docs.

File: agent/structs/intention_oss.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+package structs
+
+import (
+	"github.com/hashicorp/consul/acl"
+)
+
+func (_ *Intention) FillAuthzContext(_ *acl.AuthorizerContext, _ bool) {
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/structs/intention_oss.go]: Will do

File: agent/consul/server_test.go [hunk 7]
```diff
@@ -29,6 +29,23 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+const (
+	TestDefaultMasterToken = "d9f05e83-a7ae-47ce-839e-c0d53a68c00a"
+)
+
+func testServerACLConfig(cb func(*Config)) func(*Config) {
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/consul/server_test.go]: Will add a comment. In this case cb is going to be a pass through to testServerWithConfig and cb in that context is some function to run which can modify the server Config prior to starting the server.

File: agent/structs/intention.go [hunk 8]
```diff
@@ -247,6 +245,43 @@ func (x *Intention) Validate() error {
 	return result
 }
 
+func (ixn *Intention) CanRead(authz acl.Authorizer) bool {
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/structs/intention.go]: Will do

File: agent/consul/intention_endpoint.go (hunk 1 above)
Comment (mkcp) [reply agent/consul/intention_endpoint.go]: Makes sense -- and I agree that a refactor here is pretty big and probably out of scope of this task. If you do find yourself coming back to this function, I'd be interested in pairing with you to learn from how you'd approach the refactor!

Comment (mkeeler) [reply agent/consul/intention_endpoint.go]: @mkcp Let me know what you think of the new Intention.Apply RPC endpoint. I refactored things to do the Op switch and then offload to a op specific function. This removed much of the conditional branching in this RPC that I thought made it hard to follow.

It does duplicate a few things but IMO, I like the new way better. What do you think?

Comment (mkcp) [reply agent/consul/intention_endpoint.go]: I think the refactor works great! 💯  The control flow of the function is clear at a glance, and the names you chose help clearly delineate the branches.

**Comment by mkeeler** [comment]
I checked it out and the one failing test is a flake and unrelated to these changes (and it executes properly locally). Going to go ahead and merge this anyways.
